### ISSUE TITLE
unrunnable configuration after dumpPython with sequence containing deleted module

### ISSUE THREAD

**Issue by slava77** [cms-sw/cmssw#15792]
There is a problem with configuration produced by dumpPython for sequences containing deleted module

```
process.seqB = cms.Sequence(process.modA + process.modB)
del process.modB
```

this will run with a original python config.

If instead, config from process.dumpPython  is used we get 
`process.seqB = cms.Sequence(process.modA + process.None)`
which does not run

**Comment by Dr15Jones** [comment]
I'm not saying that shouldn't try to fix these, but it is problems like these which is why the workflow system abandoned using dumpPython many years ago.

**Comment by davidlange6** [comment]
only it didn’t abandon it.…

**Comment by Dr15Jones** [comment]
@ericvaandering My understanding was pickling allowed WMAgent to avoid use of dumpPython

**Comment by ericvaandering** [comment]
We did:

```
Samburu:WMCore ewv$ git grep dumpPython
CHANGES:  - Use pickle instead of dumpPython when serializing CMSSW PSets.  Fixes #4009.
```

Yours truly in 2012: https://github.com/dmwm/WMCore/pull/4023

**Comment by ericvaandering** [comment]
What do you mean by "3 ways"? CRAB and WMAgent at least do things almost the same. The difference is that WMAgent jobs start with the base PSet without all the imports resolved while CRAB cannot do this since there are user-level changes. So CRAB works with a pickled version on the submit node and WMAgent doesn't use pickle until the WN. But the input in CMSSW should be the same.

**Comment by davidlange6** [comment]
cmsDriver has long had two ways (one used by ~everything outside of Config/DataProcessing and another used by Confg/DataProcessing, which is aka, the pre 2012 WMAgent way) - neither being the way CRAB/WMAgent does it now (or since 2012).

**Comment by slava77** [comment]
On 9/9/16 6:05 AM, David Lange wrote:
> cmsDriver has long had two ways (one used by ~everything outside of
> Config/DataProcessing and another used by Confg/DataProcessing, which is
> aka, the pre 2012 WMAgent way) - neither being the way CRAB/WMAgent does
> it now (or since 2012).
> 

We can update the test DataProcessing workflow to use pickle.
Could someone please give some pointers (copy-pasteable example is the best)

Thanks.

**Comment by ericvaandering** [comment]
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py#L738 May be what you want. I believe we write two files, one text, one binary, to avoid issues with python strings not being able to capture every possible output of pickle.

Note this code is a bit old and should really be done with context managers and in python 2.7 you can even open two files in a single context manager.

**Comment by wddgit** [comment]
Using pickle instead of dumpPython sounds like a reasonable idea and I am not against that.

On the other hand, one thing I notice looking at the python code. If one replaces a module already attached to the Process object, there are 2 cases:

Case 1: If it is during a process load, it checks to see if the module is already on a path, endpath, or sequence that is attached to the process and raises an exception if it is.

Case 2: If it is not during a process load, it replaces the module with the new value everywhere it appears on a path, endpath or sequence attached to the process (it could be directly attached to the process or directly or indirectly included by something attached to the process).

We do not do this when deleting an attribute of the process. It would be removing instead of replacing, but it seems intuitive that if we do it for replace we should also do it for delete.

I am not sure this is worth the effort, but I could implement this. Already I am in the middle of this python code implementing Tasks. So it is fresh in my mind.

This would probably fix the problem Slava is encountering. It would fix most cases where None shows up in dumpPython output in a Sequence. Although one can find ways to work around this at the top level. For example, create a sequence at the top level, then replace or delete a module on it,  and then add the sequence to a path already attached to the process. So modifying the delete method would not be bulletproof, but would resolve all the common cases.

If we really wanted it to be bulletproof here are some other options we could use instead of or in addition to the idea above when dumpPython hits a module with a label set to the value None:

A. Raise an exception in dumpPython.
B. Write out the full module definition in the dumpPython output in the middle of the sequence.
C. Write out a dummy placeholder that would cause an exception to be raised if a module ever ran that sequence, but otherwise would not cause problems if the dumpPython output was run. (Chris suggested this as a possibility when we were discussing this)






I'll add one other thing that Chris mentioned since I do not see that Chris did not add it to this issue yet. When we encounter a module with a label that has the value None, we could

**Comment by slava77** [comment]
I would expect that "None" is ignored in parsing and actual schedule construction as a part of this issue resolution.

**Comment by Dr15Jones** [comment]
If one deletes it from the process and then later attaches a Sequence containing the deleted item you'd still get `None`. Effectively we have no way of understanding ones intent under that case since you could mean

- I didn't realize the Sequence had that module but I really don't want it anyway or
- I didn't realize the Sequence had that module and what I really want to do only works of that module is actually there and I'd like the job to fail so I can catch the mistake.

**Comment by slava77** [comment]
On 9/9/16 9:00 AM, Chris Jones wrote:
> If one deletes it from the process and then later attaches a Sequence
> containing the deleted item you'd still get |None|. Effectively we have
> no way of understanding ones intent under that case since you could mean
> 
>   * I didn't realize the Sequence had that module but I really don't
>     want it anyway or
>   * I didn't realize the Sequence had that module and what I really want
>     to do only works of that module is actually there and I'd like the
>     job to fail so I can catch the mistake.


sounds analogous to an auto-correct that can not be disabled.


On a more practical note.
A delete was called for a module object, but didn't delete the element
from the sequences.
I'd be fine to cleanup the sequences as well.
Please suggest a way to do it (preferably without making subsequences
expand).
I suppose this will need to be removed from the sequence first while the
identity is still known
and then deleted.

**Comment by wddgit** [comment]
The replace function currently used when replacing a module does the opposite. It expands the sequence at and above the level where the change was made and avoids modifying the subsequences that are inside of sequences (unless they are also attached to the process and get changed for that reason).

We could instead find all the sequences and contained sequences whether they are attached to the process or not, then replace and/or remove at the top level of each sequence only without expanding any of them.  It seems intuitive that replace and remove would have similar behaviors, although I suppose it does not have to be that way.

**Comment by wddgit** [comment]
I discussed this with Chris some more. We think it makes sense to change the behavior so that deleting a module from the process causes the module to be removed from all paths, endpaths, and Sequences that are attached to the process. It is something we could reasonably do. It makes deleting more like replacing. This might be enough to remove all the None's we currently see in the output of dumpPython. The other ways to make these None's might be sufficiently rare that they never happen in practice.

But replace does a lot of expanding of subsequences as it does it job. There are reasons replace is implemented this way. We do not want calling replace on a Path to change subsequences that could be used in some other Path (or in some other process if we are running SubProcesses). So the contents of the subsequences  are taken out of the existing subsequence and expanded into a new sequence. The side effects of actually changing the subsequence objects would cause unpredictable serious problems. So we think this deleting should expand in the same way that replace currently expands.

**Comment by slava77** [comment]
On 9/9/16 12:24 PM, W. David Dagenhart wrote:
> I discussed this with Chris some more. We think it makes sense to change
> the behavior so that deleting a module from the process causes the
> module to be removed from all paths, endpaths, and Sequences that are
> attached to the process. It is something we could reasonably do. It
> makes deleting more like replacing. This might be enough to remove all
> the None's we currently see in the output of dumpPython. The other ways
> to make these None's might be sufficiently rare that they never happen
> in practice.
> 
> But replace does a lot of expanding of subsequences as it does it job.
> There are reasons replace is implemented this way. We do not want
> calling replace on a Path to change subsequences that could be used in
> some other Path (or in some other process if we are running
> SubProcesses). So the contents of the subsequences are taken out of the
> existing subsequence and expanded into a new sequence. The side effects
> of actually changing the subsequence objects would cause unpredictable
> serious problems. So we think this deleting should expand in the same
> way that replace currently expands.


I guess if we get identical results, it is still OK.
I do not know offhand what else is going on in (mostly pat)
configuration building
that may still depend on subsequences.

Expanding subsequences makes debugging of configuration building much
more complicated
when relying on pythonDump output.

**Comment by Dr15Jones** [comment]
assign core

**Comment by slava77** [comment]
[closing]
- pickle for T0 test configs was implemented in #17635
- the task feature available since #17801 apparently improved the situation as well
